core/lib/zksync_core/src/eth_watch/event_processors/governance_upgrades.rs

```diff
@@ -0,0 +1,119 @@
+    ProtocolUpgrade, ProtocolVersionId, H256,
+};
+
+/// Listens to operation events coming from the governance contract and saves new protocol upgrade proposals to the database.
+#[derive(Debug)]
+pub struct GovernanceUpgradesEventProcessor {
+    diamond_proxy_address: Address,
+    /// Last protocol version seen. Used to skip events for already known upgrade proposals.
+    last_seen_version_id: ProtocolVersionId,
+    upgrade_proposal_signature: H256,
+}
+
+impl GovernanceUpgradesEventProcessor {
+    pub fn new(
+        diamond_proxy_address: Address,
+        last_seen_version_id: ProtocolVersionId,
+        governance_contract: &Contract,
+    ) -> Self {
+        Self {
+            diamond_proxy_address,
+            last_seen_version_id,
+            upgrade_proposal_signature: governance_contract
+                .event("TransparentOperationScheduled")
+                .expect("TransparentOperationScheduled event is missing in abi")
+                .signature(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<W: EthClient + Sync> EventProcessor<W> for GovernanceUpgradesEventProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut StorageProcessor<'_>,
+        client: &W,
+        events: Vec<Log>,
+    ) -> Result<(), Error> {
+        let mut upgrades = Vec::new();
+        for event in events
+            .into_iter()
+            .filter(|event| event.topics[0] == self.upgrade_proposal_signature)
+        {
+            let governance_operation = GovernanceOperation::try_from(event)
+                .map_err(|err| Error::LogParse(format!("{:?}", err)))?;
+            // Some calls can target other contracts than Diamond proxy, skip them.
+            for call in governance_operation
+                .calls
+                .into_iter()
+                .filter(|call| call.target == self.diamond_proxy_address)
+            {
+                let upgrade = ProtocolUpgrade::try_from(call)
+                    .map_err(|err| Error::LogParse(format!("{:?}", err)))?;
+                // Scheduler VK is not present in proposal event. It is hardcoded in verifier contract.
+                let scheduler_vk_hash = if let Some(address) = upgrade.verifier_address {
+                    Some(client.scheduler_vk_hash(address).await?)
+                } else {
+                    None
+                };
+                upgrades.push((upgrade, scheduler_vk_hash));
+            }
+        }
+
+        if upgrades.is_empty() {
+            return Ok(());
+        }
+
+        let ids_str: Vec<_> = upgrades
+            .iter()
+            .map(|(u, _)| format!("{}", u.id as u16))
+            .collect();
+        tracing::debug!("Received upgrades with ids: {}", ids_str.join(", "));
+
+        let new_upgrades: Vec<_> = upgrades
+            .into_iter()
+            .skip_while(|(v, _)| v.id as u16 <= self.last_seen_version_id as u16)
```

Comment:
should this be 'skip_while' or 'filter'? do we have a guarantee that we are getting things in order?

Reply:
Yes, web3 API guarantees us that events are received in order